File: Arrays & Hashing/0930. Binary Subarrays With Sum/0930.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int,int> prefixSum;
        prefixSum[0] = 1;
        int res = 0;
        int currSum = 0;

        for (auto& num : nums) {
            currSum += num;
            if (prefixSum.find(currSum - goal) != prefixSum.end()) {
                res += prefixSum[currSum - goal];
            }
            prefixSum[currSum]++;
        }
        return res;
    }
};

File: Arrays & Hashing/0930. Binary Subarrays With Sum/0930.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        prefix_sum = defaultdict(int)
        prefix_sum[0] = 1
        curr_sum = res = 0

        for num in nums:
            curr_sum += num
            res += prefix_sum[curr_sum - goal]
            prefix_sum[curr_sum] += 1
        
        return res